### PR TITLE
Implement Tray Icons

### PR DESCRIPTION
**Pull request by 4e554c4c**
There are still issues with this commit, mostly in dealing with `xembedsniproxy`. It creates dummy windows for placing XEmbed items which sway renders even though it shouldn't. Also `xembedsniproxy` is a super buggy program, so there still may be undiscovered problems with it.

This commit implements the StatusNotifierItem protocol, and enables
swaybar to show tray icons. It also uses `xembedsniproxy` in order to
communicate with xembed applications.
The tray is completely optional, and can be disabled on compile time
with the `enable-tray` option. Or on runtime with the bar config option
`tray_output none`.

Overview of changes:
In swaybar very little is changed outside the tray subfolder except
that all events are now polled in `event_loop.c`, this creates no
functional difference.

Six bar configuration options were added, these are detailed in
sway-bar(5)

The tray subfolder is where all protocol implementation takes place and
is organised as follows:

tray/sni_watcher.c:
  This file contains the StatusNotifierWatcher. It keeps track of
  items and hosts and reports when they come or go.
tray/tray.c
  This file contains the StatusNotifierHost. It keeps track of
  sway's version of the items and represents the tray itself.
tray/sni.c
  This file contains the StatusNotifierItem struct and all
  communication with individual items.
tray/icon.c
  This file implements the icon theme protocol. It allows for
  finding icons by name, rather than by pixmap.
tray/dbus.c
  This file allows for asynchronous DBus communication.

Note: Package maintainers sholud add xembedsniproxy and libdbus as optional dependencies of sway.

See #986 #343

### CODE REVIEW

**Comment by SirCmpwn** [comment]
One commit ;_;

File: swaybar/bar.c [hunk 1]
```diff
@@ -92,6 +120,30 @@ static void mouse_button_notify(struct window *window, int x, int y,
...
+			} else if (button == swaybar.config->context_button) {
+				sni_context_menu(item->ref, x, y);
+			} else if (button == swaybar.config->secondary_button) {
+				sni_secondary(item->ref, x, y);
+			}
+			break;
+		}
+		tray_width -= icon_width;
+	}
+#endif
```
Comment (SirCmpwn) [inline swaybar/bar.c]: Can you move this into another file?

File: swaybar/bar.c [hunk 2]
```diff
@@ -50,18 +57,39 @@ static void spawn_status_cmd_proc(struct bar *bar) {
...
+			if (pid == 0) {
+				execlp("xembedsniproxy", "xembedsniproxy", NULL);
+				_exit(EXIT_FAILURE);
+			}
+			waitpid(pid, &wstatus, 0);
+		} while (!WIFEXITED(wstatus));
+		_exit(EXIT_FAILURE);
+	}
+}
+#endif
```
Comment (SirCmpwn) [inline swaybar/bar.c]: Consider moving this as well, just have a tray upkeep function that checks this and refires it if not running

File: swaybar/bar.c [hunk 3]
```diff
@@ -178,23 +233,54 @@ void bar_setup(struct bar *bar, const char *socket_path, const char *bar_id) {
...
+		/* Start the SNI watcher */
+		init_sni_watcher();
+
+		/* Start the SNI host */
+		init_tray();
+
+		/* Start xembedsniproxy */
+		spawn_xembed_sni_proxy();
+	}
+#endif
```
Comment (SirCmpwn) [inline swaybar/bar.c]: This one too, just make it a simple

```c
#ifdef ENABLE_TRAY
    init_tray();
#endif
```

Then make init_tray do all this work. Try to limit the tray code in this file to simple calls to another function.

File: swaybar/render.c [hunk 4]
```diff
@@ -297,6 +301,72 @@ void render(struct output *output, struct config *config, struct status_line *li
...
+
+		item->dirty = false;
+	}
+
+
+	if (tray_width != window->width * window->scale) {
+		tray_width -= tray_padding;
+	}
+
+no_tray:
```
Comment (SirCmpwn) [inline swaybar/render.c]: Ditto here. Move it all into a separate function in another file.

**Comment by SirCmpwn** [comment]
Problems I noticed while testing:

* SNI works quite well, nice job. The exception is that I can't seem to select context menu items that are occluded by other windows. Does the compositor need to do something special to let them keep focus? (i.e. open konversation's context menu and try to use any option that is displayed on top of another window, such as the konversation main window). Clicking outside of the context menu also doesn't seem to close it, and should. Making sure the context menu windows are getting focus should fix both of these issues.
* xembed doesn't work at all for me. The issues with it need to be ironed out or it needs to be removed before I'm comfortable merging this.

**Comment by SirCmpwn** [comment]
I also pushed a commit fixing up the cairo blending operator so that partially transparent icons wouldn't overwrite the background

**Comment by SirCmpwn** [comment]
Status?

**Comment by 4e554c4c** [comment]
Still working on getting tray windows working, pretty much everything else is in order.

**Comment by SirCmpwn** [comment]
Should I hold off on rc1 for this?

**Comment by 4e554c4c** [comment]
I'd say not, unless you really want it there. I can't work on this 100% right now, so it might be good to delay things. Also I'm not even sure if it'd be good to release sni support without xembed included. People who don't know the difference will file reports.

**Comment by SirCmpwn** [comment]
I'm okay with shipping SNI support alone, we already get plenty of stupid support requets and I doubt this will add much. I'll have a look at this in the morning and see if I can clean it up and still feel good about it.

**Comment by SirCmpwn** [comment]
Spent some time poking on this. Results: focus is working good enough for now. Also I figured out what was wrong witht he firefox address bar dropdown and fixed that. Going to include this in the next release and maybe iterate on it a bit during rc's to fix the edge cases.

**Comment by 4e554c4c** [comment]
Thanks!

**Comment by 4e554c4c** [comment]
I'll continue work on xembed, hopefully we can have that in the next release.

**Comment by nemanjan00** [comment]
Any progress?

**Comment by SirCmpwn** [comment]
Yep.

**Pull request merged**